README.md
# Yet another hook library

Hook library for x86-32, Windows

no dependencies

---

* Work with **cdecl**, **stdcall**, **thiscall** calling conventions
* Pass all parameters by reference to hook function
* Easy ignore original function call


# Examples

### Change params in hook function

```cpp
int _declspec(noinline) sum(int a, int b) {
	return a + b;
}

void hooked_sum(int& a, int& b) {
	a = 10;
	b = 20;
}

int main() {
	hook test(sum, hooked_sum);
	std::cout << sum(1, 2);

	return 0;
}
```
```
Output:
30
```

### Ignore original function
```cpp
int _declspec(noinline) sum(int a, int b) {
	return a + b;
}

int hooked_sum(int& a, int& b) {
	
	//If hook function have "void" return type, original function will be called
	//Otherwise no	
	return 42;
}

int main() {
	hook test(sum, hooked_sum);
	std::cout << sum(1, 2);

	return 0;
}
```
```
Output:
42
```


### Work well with compound types
```cpp
void _declspec(noinline) print(std::string message, int number) {
	std::cout << message << " and number " << number;
}

void hooked_print(std::string& message, int& number) {
	message += "_hooked";
	number = 1337;
}

int main() {
	hook test(print, hooked_print);
	
	print("hello_world", 640);

	return 0;
}
```
```
Output:
hello_world_hooked and number 1337
```


### Member function support
```cpp
struct dummy {
	void func(bool a) {
		std::cout << foo << " " << bar << " " << a;
	}
	
	std::string foo;
	int bar;
};

void hooked_dummy_func(dummy*& obj, bool& a) {
	obj->foo = "example hooked text";
	a = false;
}

int main() {
	hook test(&dummy::func, hooked_dummy_func);
	
	dummy tmp;
	tmp.foo = "Hello, world!!!";
	tmp.bar = 101010101;
	
	tmp.func(true);
	
	return 0;
}
```
```
Output:
example hooked text 101010101 0
```


hde32/NEWS
version 0.28.01 [2009.03.09]
 + updated to HDE32 0.28

version 0.27.01 [2009.01.02]
 + fixed bug: error setting 0x800000 flag to "hde32s.flags" when no prefixes

version 0.25.01 [2008.10.08]
 + updated to HDE32 0.25

version 0.24.01 [2008.09.11]
 + updated to HDE32 0.24

version 0.20.01 [2008.09.06]
 + updated to HDE32 0.20

version 0.19.01 [2008.08.28]
 + updated to HDE32 0.19

version 0.15.01 [2008.08.16]
 + updated to HDE32 0.15
 + new version format: x.xx.yy, x.xx - base HDE32 version, yy - HDE32C version

version 0.02 [2008.08.09] (new version format: 0.13.02)
 + added readme.txt and license.txt

version 0.01 [2008.08.04] (new version format: 0.13.01)
 + first release


hook.h
#pragma once
#include <array>
#include <vector>
#include <memory>

#include "xbyak/xbyak.h"
#include "hde32/include/hde32.h"
#include "hook_function_traits.hpp"

namespace detail {

constexpr auto JUMP_OPCODE_SIZE = 5;
constexpr auto PARAM_SIZE = 4;

/*
	By unknown reason, C++ does not support converting a pointer to a member function to an arbitrary pointer.
	So, this "semantic sledgehammer" used for fix that
*/
template<typename Out, typename In>
constexpr Out force_cast(In in)
{
	union
	{
		In  in;
		Out out;
	}
	u = { in };

	return u.out;
};

template<typename Func1, typename Func2>
constexpr bool is_same_custom_v = std::is_same_v<get_function_args_t<Func1>, get_function_args_t<Func2>> &&
	(std::is_same_v<get_function_return_type_t<Func1>, get_function_return_type_t<Func2>> 
		|| std::is_void_v<get_function_return_type_t<Func1>>);


namespace memory_protect {
	enum class protection {
		R,
		RW,
		RE,
		RWE
	};

	auto constexpr get_system_protection_constant(protection mode) noexcept {
#if defined(_WIN32)
		constexpr auto DEFAULT_CONSTANT = PAGE_EXECUTE_READWRITE;

		if (mode == protection::R)
			return PAGE_READONLY;
		else if (mode == protection::RW)
			return PAGE_READWRITE;
		else if (mode == protection::RE)
			return PAGE_EXECUTE_READ;
		else if (mode == protection::RWE)
			return PAGE_EXECUTE_READWRITE;
		else
			return DEFAULT_CONSTANT;
#elif defined(__linux__)
		auto DEFAULT_CONSTANT = PROT_READ | PROT_WRITE | PROT_EXEC;

		if (mode == protection::R)
			return PROT_READ;
		else if (mode == protection::RW)
			return PROT_READ | PROT_WRITE;
		else if (mode == protection::RE)
			return PROT_READ | PROT_EXEC;
		else if (mode == protection::RWE)
			return PROT_READ | PROT_WRITE | PROT_EXEC;
		else
			return DEFAULT_CONSTANT;
#endif
	}

	template<typename T>
	protection constexpr convert_system_protection_constant(T constant) noexcept {
		const protection DEFAULT_PROTECTION = protection::RWE;
#if defined(_WIN32)
		if (constant == PAGE_READONLY)
			return protection::R;
		else if (constant == PAGE_READWRITE)
			return protection::RW;
		else if (constant == PAGE_EXECUTE_READ)
			return protection::RE;
		else if (constant == PAGE_EXECUTE_READWRITE)
			return protection::RWE;
		else
			return DEFAULT_PROTECTION;
#elif defined(__linux__)
		if (constant == PROT_READ)
			return protection::R;
		else if (constant == PROT_READ | PROT_WRITE)
			return protection::RW;
		else if (constant == PROT_READ | PROT_EXEC)
			return protection::RE;
		else if (constant == PROT_READ | PROT_WRITE | PROT_EXEC)
			return protection::RWE;
		else
			return DEFAULT_PROTECTION;
#endif
	}

	template<typename AddrType>
	protection set_protection(AddrType address, size_t size, protection mode) {
#if defined(_WIN32)
		const DWORD prot = get_system_protection_constant(mode);
		DWORD prev_prot;
		
		VirtualProtect(force_cast<LPVOID>(address), size, prot, &prev_prot);

		return convert_system_protection_constant(prev_prot);
#elif defined(__linux__)
		// linux unprotect implementation... where?

		/*
			Linux doesn't have any syscall for retrieve protection of memory region
			So, at default, we use READ WRITE EXECUTE
		*/
		return convert_system_protection_constant(PROT_READ | PROT_WRITE | PROT_EXEC);
#endif
	}

	template<typename AddrType>
	class unprotect_guard {
	public:
		unprotect_guard(AddrType address, size_t size): address(address), size(size) {
			prev_protection = set_protection(address, size, protection::RWE);
		}

		~unprotect_guard() {
			set_protection(address, size, prev_protection);
		}

	private:
		AddrType address;
		size_t size;
		protection prev_protection;
	};
};

class trampouline_generator {
public:
	trampouline_generator() : code(std::make_shared<Xbyak::CodeGenerator>()) {};

	template<size_t Size>
	void push_references_to_params_in_stack(const std::array<size_t, Size>& params_size) {
		using namespace Xbyak::util;

		auto params_count = Size;

		for (size_t i = 0; i < params_count; i++) {
			auto push_offset = sum_first_until_n(params_size, params_count - (i+1));
			push_offset += (i+1) * PARAM_SIZE;

			code->lea(eax, dword[esp + push_offset]);
			code->push(eax);
		}
	}

	template<size_t Size>
	void push_references_to_params_and_ecx_in_stack(const std::array<size_t, Size>& params_size) {
		using namespace Xbyak::util;

		code->push(ecx);

		constexpr auto params_count = Size;

		for (size_t i = 0; i < params_count; i++) {
			auto push_offset = sum_first_until_n(params_size, params_count - (i + 1));
			push_offset += (i + 2) * PARAM_SIZE;

			code->lea(eax, dword[esp + push_offset]);
			code->push(eax);
		}

		code->lea(eax, dword[esp + params_size.size() * PARAM_SIZE]);
		code->push(eax);
	}
	
	void copy_params_in_stack(size_t params_size) {
		using namespace Xbyak::util;

		for (size_t i = 0; i < params_size; i += PARAM_SIZE) {
			code->push(dword [esp + params_size]);
		}
	}

	template<typename Func>
	void call_function(Func function) {
		using namespace Xbyak::util;

		code->call(force_cast<void *>(function));
		
		if constexpr (get_calling_convention_v<Func> == cc::cdecl_) {
			code->add(esp, params_size_v<std::remove_pointer_t<Func>>);
		}
	}

	template<typename Func, typename Obj>
	void call_method(Func method, Obj object) {
		using namespace Xbyak::util;

		code->mov(ecx, force_cast<uint32_t>(object));
		code->call(force_cast<void *>(method));
	}

	template<typename Func> 
	void return_from() {
		if constexpr (get_calling_convention_v<Func> == cc::stdcall_ || 
				get_calling_convention_v<Func> == cc::thiscall_) {
			code->ret(params_size_v<Func>);
		} 
		else {
			code->ret();
		}
	}

	void save_registers() {
		code->pushad();
	}

	void restore_registers() {
		code->popad();
	}

	void push_ecx_by_reference() {
		using namespace Xbyak::util;

		code->lea(eax, dword [esp - PARAM_SIZE*2]);
		code->push(eax);
		code->push(ecx);
		code->add(esp, PARAM_SIZE);
	}

	void pop_ecx() {
		using namespace Xbyak::util;

		code->pop(ecx);
	}

	std::shared_ptr<Xbyak::CodeGenerator> get_code() {
		code->ready();

		return code;
	}

private:
	template<typename T>
	static constexpr auto sum_first_until_n(T& container, size_t n) {
		typename T::value_type sum = 0;
		for (auto it = container.cbegin(); it != container.cbegin() + n; it++) {
			sum += *it;
		}

		return sum;
	}

	std::shared_ptr<Xbyak::CodeGenerator> code;
};

}

template<typename HookeeType, typename HookType>
class hook {
public:
	hook(HookeeType source, HookType destination) : hookee_address(source), hook_address(destination),
		original_instructions(
			get_full_instructions_at_least_n_bytes(detail::force_cast<uint8_t*>(hookee_address), detail::JUMP_OPCODE_SIZE)) {
		assert_check_function_types();

		generate_trampouline();
		set_jump_to_trampouline();
	};

	~hook() {
		restore_original_instructions();
	}

	hook(const hook&) = delete;
	hook(hook&&) = delete;
	hook& operator=(const hook&) = delete;
	hook& operator=(hook&&) = delete;

private:
	static void constexpr assert_check_function_types() {
		static_assert(get_calling_convention_v<HookeeType> != cc::unknown_ ||
					  get_calling_convention_v<HookType> != cc::unknown_,
			"Unknown calling convention");
		
		if constexpr (get_calling_convention_v<HookeeType> == cc::thiscall_) {
			static_assert(detail::is_same_custom_v<remove_func_pointer_t<HookType>,
				add_reference_to_args_t<
					convert_member_to_free_func_t<HookeeType>>>,
				"For member function, hook function must get params by reference and have a first argument \"*this\"");
		}
		else {
			static_assert(detail::is_same_custom_v<
										remove_func_pointer_t<HookType>,
				add_reference_to_args_t<remove_func_pointer_t<HookeeType>>>,
				"Hook function must have same signature and get params by reference");
		}
	};

	static std::vector<uint8_t>
	get_full_instructions_at_least_n_bytes(const uint8_t* function, size_t n) {
		std::vector<uint8_t> instructions;

		while (instructions.size() < n) {
			hde32s instruction;

			hde32_disasm(function, &instruction);
			std::copy(function, function + instruction.len, std::back_inserter(instructions));
			function += instruction.len;
		}

		return instructions;
	}

	void generate_trampouline() {
		detail::trampouline_generator code_gen;

		using rp_HookeeType = remove_func_pointer_t<HookeeType>;
		using rp_HookType = remove_func_pointer_t<HookType>;

		constexpr auto is_thiscall = get_calling_convention_v<rp_HookeeType> == cc::thiscall_;
		constexpr auto call_original = std::is_void_v<get_function_return_type_t<rp_HookType>>;
		constexpr auto hookee_params_array = params_size_as_array_v<rp_HookeeType>;
		constexpr auto hookee_params_size = params_size_v<rp_HookeeType>;

		if constexpr (is_thiscall) {
			code_gen.push_references_to_params_and_ecx_in_stack(hookee_params_array);
		}
		else {
			code_gen.push_references_to_params_in_stack(hookee_params_array);
		}
		code_gen.call_function(hook_address);
		if constexpr (is_thiscall) {
			code_gen.pop_ecx();
		}
		
		if constexpr (call_original) {
			code_gen.save_registers();
			code_gen.call_method(&hook<HookeeType, HookType>::restore_original_instructions, this);
			code_gen.restore_registers();

			code_gen.copy_params_in_stack(hookee_params_size);
			code_gen.call_function(hookee_address);

			code_gen.save_registers();
			code_gen.call_method(&hook<HookeeType, HookType>::set_jump_to_trampouline, this);
			code_gen.restore_registers();
		}

		code_gen.return_from<rp_HookeeType>();

		trampouline_code = code_gen.get_code();
	}

	void set_jump_to_trampouline() {
		using namespace detail;
		memory_protect::unprotect_guard guard(hookee_address, 5);

		constexpr unsigned char JMP_REL32_BYTECODE = 0xE9;
		constexpr unsigned char NOP_BYTECODE = 0x90;
		std::array<uint8_t, JUMP_OPCODE_SIZE> jump_opcode;
		
		jump_opcode[0] = JMP_REL32_BYTECODE;

		uint32_t jump_offset = force_cast<uint32_t>(trampouline_code->getCode()) - 
			force_cast<uint32_t>(hookee_address) - JUMP_OPCODE_SIZE;
		*(uint32_t*)&jump_opcode[1] = jump_offset;

		std::copy(jump_opcode.begin(), jump_opcode.end(),
			force_cast<uint8_t*>(hookee_address));
			
		if (original_instructions.size() > JUMP_OPCODE_SIZE) {
			std::fill(force_cast<uint8_t*>(hookee_address) + JUMP_OPCODE_SIZE,
				force_cast<uint8_t*>(hookee_address) + original_instructions.size(), NOP_BYTECODE);
		}
	}

	void restore_original_instructions() {
		detail::memory_protect::unprotect_guard guard(hookee_address, 5);

		std::copy(original_instructions.begin(), original_instructions.end(), 
			detail::force_cast<uint8_t*>(hookee_address));
	}

	HookeeType hookee_address;
	HookType hook_address;
	std::vector<uint8_t> original_instructions;
	std::shared_ptr<Xbyak::CodeGenerator> trampouline_code;
};

hde32/include/hde32.h
/*
* Hacker Disassembler Engine 32
* Copyright (c) 2006-2009, Vyacheslav Patkov.
* All rights reserved.
*
* hde32.h: C/C++ header file
*
*/

#ifndef _HDE32_H_
#define _HDE32_H_

#include <stdint.h>

#define F_MODRM         0x00000001
#define F_SIB           0x00000002
#define F_IMM8          0x00000004
#define F_IMM16         0x00000008
#define F_IMM32         0x00000010
#define F_DISP8         0x00000020
#define F_DISP16        0x00000040
#define F_DISP32        0x00000080
#define F_RELATIVE      0x00000100
#define F_2IMM16        0x00000800
#define F_ERROR         0x00001000
#define F_ERROR_OPCODE  0x00002000
#define F_ERROR_LENGTH  0x00004000
#define F_ERROR_LOCK    0x00008000
#define F_ERROR_OPERAND 0x00010000
#define F_PREFIX_REPNZ  0x01000000
#define F_PREFIX_REPX   0x02000000
#define F_PREFIX_REP    0x03000000
#define F_PREFIX_66     0x04000000
#define F_PREFIX_67     0x08000000
#define F_PREFIX_LOCK   0x10000000
#define F_PREFIX_SEG    0x20000000
#define F_PREFIX_ANY    0x3f000000

#define PREFIX_SEGMENT_CS   0x2e
#define PREFIX_SEGMENT_SS   0x36
#define PREFIX_SEGMENT_DS   0x3e
#define PREFIX_SEGMENT_ES   0x26
#define PREFIX_SEGMENT_FS   0x64
#define PREFIX_SEGMENT_GS   0x65
#define PREFIX_LOCK         0xf0
#define PREFIX_REPNZ        0xf2
#define PREFIX_REPX         0xf3
#define PREFIX_OPERAND_SIZE 0x66
#define PREFIX_ADDRESS_SIZE 0x67

#pragma pack(push,1)

typedef struct {
	uint8_t len;
	uint8_t p_rep;
	uint8_t p_lock;
	uint8_t p_seg;
	uint8_t p_66;
	uint8_t p_67;
	uint8_t opcode;
	uint8_t opcode2;
	uint8_t modrm;
	uint8_t modrm_mod;
	uint8_t modrm_reg;
	uint8_t modrm_rm;
	uint8_t sib;
	uint8_t sib_scale;
	uint8_t sib_index;
	uint8_t sib_base;
	union {
		uint8_t imm8;
		uint16_t imm16;
		uint32_t imm32;
	} imm;
	union {
		uint8_t disp8;
		uint16_t disp16;
		uint32_t disp32;
	} disp;
	uint32_t flags;
} hde32s;

#pragma pack(pop)

#ifdef __cplusplus
extern "C" {
#endif

/* __cdecl */
unsigned int hde32_disasm(const void *code, hde32s *hs);

#ifdef __cplusplus
}
#endif

#endif /* _HDE32_H_ */


hde32/src/hde32.cpp
/*
* Hacker Disassembler Engine 32 C
* Copyright (c) 2008-2009, Vyacheslav Patkov.
* All rights reserved.
*
*/

#include <stdint.h>
#include <string.h>
#include "../include/hde32.h"
#include "../include/table32.h"

#ifdef _MSC_VER
#pragma warning(disable:4701)
#endif

unsigned int hde32_disasm(const void *code, hde32s *hs)
{
	uint8_t x, c, *p = (uint8_t*)code, cflags, opcode, pref = 0;
	uint8_t* ht = hde32_table, m_mod, m_reg, m_rm, disp_size = 0;

	memset(hs,0,sizeof(hde32s));

	for (x = 16; x; x--)
		switch (c = *p++) {
		case 0xf3:
			hs->p_rep = c;
			pref |= PRE_F3;
			break;
		case 0xf2:
			hs->p_rep = c;
			pref |= PRE_F2;
			break;
		case 0xf0:
			hs->p_lock = c;
			pref |= PRE_LOCK;
			break;
		case 0x26: case 0x2e: case 0x36:
		case 0x3e: case 0x64: case 0x65:
			hs->p_seg = c;
			pref |= PRE_SEG;
			break;
		case 0x66:
			hs->p_66 = c;
			pref |= PRE_66;
			break;
		case 0x67:
			hs->p_67 = c;
			pref |= PRE_67;
			break;
		default:
			goto pref_done;
	}
pref_done:

	hs->flags = (uint32_t)pref << 23;

	if (!pref)
		pref |= PRE_NONE;

	if ((hs->opcode = c) == 0x0f) {
		hs->opcode2 = c = *p++;
		ht += DELTA_OPCODES;
	} else if (c >= 0xa0 && c <= 0xa3) {
		if (pref & PRE_67)
			pref |= PRE_66;
		else
			pref &= ~PRE_66;
	}

	opcode = c;
	cflags = ht[ht[opcode / 4] + (opcode % 4)];

	if (cflags == C_ERROR) {
		hs->flags |= F_ERROR | F_ERROR_OPCODE;
		cflags = 0;
		if ((opcode & -3) == 0x24)
			cflags++;
	}

	x = 0;
	if (cflags & C_GROUP) {
		uint16_t t;
		t = *(uint16_t*)(ht + (cflags & 0x7f));
		cflags = (uint8_t)t;
		x = (uint8_t)(t >> 8);
	}

	if (hs->opcode2) {
		ht = hde32_table + DELTA_PREFIXES;
		if (ht[ht[opcode / 4] + (opcode % 4)] & pref)
			hs->flags |= F_ERROR | F_ERROR_OPCODE;
	}

	if (cflags & C_MODRM) {
		hs->flags |= F_MODRM;
		hs->modrm = c = *p++;
		hs->modrm_mod = m_mod = c >> 6;
		hs->modrm_rm = m_rm = c & 7;
		hs->modrm_reg = m_reg = (c & 0x3f) >> 3;

		if (x && ((x << m_reg) & 0x80))
			hs->flags |= F_ERROR | F_ERROR_OPCODE;

		if (!hs->opcode2 && opcode >= 0xd9 && opcode <= 0xdf) {
			uint8_t t = opcode - 0xd9;
			if (m_mod == 3) {
				ht = hde32_table + DELTA_FPU_MODRM + t*8;
				t = ht[m_reg] << m_rm;
			} else {
				ht = hde32_table + DELTA_FPU_REG;
				t = ht[t] << m_reg;
			}
			if (t & 0x80)
				hs->flags |= F_ERROR | F_ERROR_OPCODE;
		}

		if (pref & PRE_LOCK) {
			if (m_mod == 3) {
				hs->flags |= F_ERROR | F_ERROR_LOCK;
			} else {
				uint8_t* table_end, op = opcode;
				if (hs->opcode2) {
					ht = hde32_table + DELTA_OP2_LOCK_OK;
					table_end = ht + DELTA_OP_ONLY_MEM - DELTA_OP2_LOCK_OK;
				} else {
					ht = hde32_table + DELTA_OP_LOCK_OK;
					table_end = ht + DELTA_OP2_LOCK_OK - DELTA_OP_LOCK_OK;
					op &= -2;
				}
				for (; ht != table_end; ht++)
					if (*ht++ == op) {
						if (!((*ht << m_reg) & 0x80))
							goto no_lock_error;
						else
							break;
					}
					hs->flags |= F_ERROR | F_ERROR_LOCK;
no_lock_error:
					;
			}
		}

		if (hs->opcode2) {
			switch (opcode) {
			case 0x20: case 0x22:
				m_mod = 3;
				if (m_reg > 4 || m_reg == 1)
					goto error_operand;
				else
					goto no_error_operand;
			case 0x21: case 0x23:
				m_mod = 3;
				if (m_reg == 4 || m_reg == 5)
					goto error_operand;
				else
					goto no_error_operand;
			}
		} else {
			switch (opcode) {
			case 0x8c:
				if (m_reg > 5)
					goto error_operand;
				else
					goto no_error_operand;
			case 0x8e:
				if (m_reg == 1 || m_reg > 5)
					goto error_operand;
				else
					goto no_error_operand;
			}
		}

		if (m_mod == 3) {
			uint8_t* table_end;
			if (hs->opcode2) {
				ht = hde32_table + DELTA_OP2_ONLY_MEM;
				table_end = ht + sizeof(hde32_table) - DELTA_OP2_ONLY_MEM;
			} else {
				ht = hde32_table + DELTA_OP_ONLY_MEM;
				table_end = ht + DELTA_OP2_ONLY_MEM - DELTA_OP_ONLY_MEM;
			}
			for (; ht != table_end; ht += 2)
				if (*ht++ == opcode) {
					if ((*ht++ & pref) && !((*ht << m_reg) & 0x80))
						goto error_operand;
					else
						break;
				}
				goto no_error_operand;
		} else if (hs->opcode2) {
			switch (opcode) {
			case 0x50: case 0xd7: case 0xf7:
				if (pref & (PRE_NONE | PRE_66))
					goto error_operand;
				break;
			case 0xd6:
				if (pref & (PRE_F2 | PRE_F3))
					goto error_operand;
				break;
			case 0xc5:
				goto error_operand;
			}
			goto no_error_operand;
		} else
			goto no_error_operand;

error_operand:
		hs->flags |= F_ERROR | F_ERROR_OPERAND;
no_error_operand:

		c = *p++;
		if (m_reg <= 1) {
			if (opcode == 0xf6)
				cflags |= C_IMM8;
			else if (opcode == 0xf7)
				cflags |= C_IMM_P66;
		}

		switch (m_mod) {
		case 0:
			if (pref & PRE_67) {
				if (m_rm == 6)
					disp_size = 2;
			} else
				if (m_rm == 5)
					disp_size = 4;
			break;
		case 1:
			disp_size = 1;
			break;
		case 2:
			disp_size = 2;
			if (!(pref & PRE_67))
				disp_size <<= 1;
			break;
		}

		if (m_mod != 3 && m_rm == 4 && !(pref & PRE_67)) {
			hs->flags |= F_SIB;
			p++;
			hs->sib = c;
			hs->sib_scale = c >> 6;
			hs->sib_index = (c & 0x3f) >> 3;
			if ((hs->sib_base = c & 7) == 5 && !(m_mod & 1))
				disp_size = 4;
		}

		p--;
		switch (disp_size) {
		case 1:
			hs->flags |= F_DISP8;
			hs->disp.disp8 = *p;
			break;
		case 2:
			hs->flags |= F_DISP16;
			hs->disp.disp16 = *(uint16_t*)p;
			break;
		case 4:
			hs->flags |= F_DISP32;
			hs->disp.disp32 = *(uint32_t*)p;
			break;
		}
		p += disp_size;
	} else if (pref & PRE_LOCK)
		hs->flags |= F_ERROR | F_ERROR_LOCK;

	if (cflags & C_IMM_P66) {
		if (cflags & C_REL32) {
			if (pref & PRE_66) {
				hs->flags |= F_IMM16 | F_RELATIVE;
				hs->imm.imm16 = *(uint16_t*)p;
				p += 2;
				goto disasm_done;
			}
			goto rel32_ok;
		}
		if (pref & PRE_66) {
			hs->flags |= F_IMM16;
			hs->imm.imm16 = *(uint16_t*)p;
			p += 2;
		} else {
			hs->flags |= F_IMM32;
			hs->imm.imm32 = *(uint32_t*)p;
			p += 4;
		}
	}

	if (cflags & C_IMM16) {
		if (hs->flags & F_IMM32) {
			hs->flags |= F_IMM16;
			hs->disp.disp16 = *(uint16_t*)p;
		} else if (hs->flags & F_IMM16) {
			hs->flags |= F_2IMM16;
			hs->disp.disp16 = *(uint16_t*)p;
		} else {
			hs->flags |= F_IMM16;
			hs->imm.imm16 = *(uint16_t*)p;
		}
		p += 2;
	}
	if (cflags & C_IMM8) {
		hs->flags |= F_IMM8;
		hs->imm.imm8 = *p++;
	}

	if (cflags & C_REL32) {
rel32_ok:
		hs->flags |= F_IMM32 | F_RELATIVE;
		hs->imm.imm32 = *(uint32_t*)p;
		p += 4;
	} else if (cflags & C_REL8) {
		hs->flags |= F_IMM8 | F_RELATIVE;
		hs->imm.imm8 = *p++;
	}

disasm_done:

	if ((hs->len = (uint8_t)(p-(uint8_t*)code)) > 15) {
		hs->flags |= F_ERROR | F_ERROR_LENGTH;
		hs->len = 15;
	}

	return (unsigned int)hs->len;
}
